sql/plan/resolved_table.go

```diff
@@ -0,0 +1,110 @@
+package plan
+
+import (
+	"fmt"
+	"io"
+
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+)
+
+// ResolvedTable represents a resolved SQL Table.
+type ResolvedTable struct {
+	sql.TableV2
+	name string
+}
+
+var _ sql.Node = (*ResolvedTable)(nil)
+
+// NewResolvedTable creates a new instance of ResolvedTable.
+func NewResolvedTable(name string, table sql.TableV2) *ResolvedTable {
+	return &ResolvedTable{table, name}
+}
+
+// Name implements the Nameable interface.
+func (t *ResolvedTable) Name() string { return t.name }
+
+// Resolved implements the Resolvable interface.
+func (*ResolvedTable) Resolved() bool {
+	return false
+}
+
+// String implements the Stringer inteface.
+func (t *ResolvedTable) String() string {
+	p := sql.NewTreePrinter()
+	_ = p.WriteNode("Table(%s)", t.name)
+	var schema = make([]string, len(t.Schema()))
+	for i, col := range t.Schema() {
+		schema[i] = fmt.Sprintf(
+			"Column(%s, %s, nullable=%v)",
+			col.Name,
+			col.Type.Type().String(),
+			col.Nullable,
+		)
+	}
+	_ = p.WriteChildren(schema...)
+	return p.String()
+}
+
+// Children implements the Node interface.
+func (*ResolvedTable) Children() []sql.Node {
+	return []sql.Node{}
+}
+
+// RowIter implements the RowIter interface.
+func (t *ResolvedTable) RowIter(ctx *sql.Context) (sql.RowIter, error) {
+	return &tableIter{table: t.TableV2, partitions: t.TableV2.Partitions(ctx)}, nil
+}
+
+// TransformUp implements the Transformable interface.
+func (t *ResolvedTable) TransformUp(f sql.TransformNodeFunc) (sql.Node, error) {
+	return f(NewResolvedTable(t.name, t.TableV2))
+}
+
+// TransformExpressionsUp implements the Transformable interface.
+func (t *ResolvedTable) TransformExpressionsUp(f sql.TransformExprFunc) (sql.Node, error) {
+	return t, nil
+}
+
+type tableIter struct {
+	table      sql.TableV2
+	partitions sql.PartitionIter
+	partition  sql.Partition
+	rows       sql.RowIter
+}
+
+func (i *tableIter) Next() (sql.Row, error) {
+	if i.partition == nil {
+		partition, err := i.partitions.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		i.partition = partition
+	}
+
+	if i.rows == nil {
+		rows, err := i.table.PartitionRows(i.partition)
+		if err != nil {
+			return nil, err
+		}
+
+		i.rows = rows
+	}
+
+	row, err := i.rows.Next()
+	if err != nil && err == io.EOF {
+		i.partition = nil
+		if err := i.rows.Close(); err != nil {
+			return nil, err
+		}
+
+		i.rows = nil
+		return i.Next()
```

Comment:
tail recursion is fine, I just thought about `goto` ;)

Reply:
Is this a case where is not considered a "bad practice" use `goto`?

Reply:
No, do not use `goto`. Tail recursion is fast and we won't have huge amount of partitions (I suppose). Personally I just like `goto` - it's fast ;)